lib/IRGen/GenCall.cpp

```diff
@@ -1515,13 +1705,323 @@ void irgen::extractScalarResults(IRGenFunction &IGF, llvm::Type *bodyType,
+    }
+    // At this point, llArgs contains the arguments that are being passed along
+    // via the async context.  We can safely drop them on the floor.
+    (void)llArgs.claimAll();
+    // TODO: Validation: we should be able to check that the contents of llArgs
+    //       matches what is expected by the layout.
+  }
+  Address getErrorSlot(SILType errorType) override {
+    auto layout = getAsyncContextLayout();
+    auto errorLayout = layout.getErrorLayout();
+    return errorLayout.project(IGF, context, /*offsets*/ llvm::None);
+  };
+};
+
+} // end anonymous namespace
+
+std::unique_ptr<CallEmission> irgen::getCallEmission(IRGenFunction &IGF,
+                                                     llvm::Value *selfValue,
+                                                     Callee &&callee) {
+  if (callee.getOrigFunctionType()->isAsync()) {
+    return std::make_unique<AsyncCallEmission>(IGF, selfValue,
+                                               std::move(callee));
+  } else {
+    return std::make_unique<SyncCallEmission>(IGF, selfValue,
+                                              std::move(callee));
+  }
+}
+
 /// Emit the unsubstituted result of this call into the given explosion.
 /// The unsubstituted result must be naturally returned directly.
 void CallEmission::emitToUnmappedExplosion(Explosion &out) {
+  assert(state == State::Emitting);
   assert(LastArgWritten == 0 && "emitting unnaturally to explosion");
 
   auto call = emitCallSite();
 
+  if (getCallee().getOrigFunctionType()->isAsync()) {
+    SILFunctionConventions fnConv(getCallee().getOrigFunctionType(),
+                                  IGF.getSILModule());
+    auto resultType =
+        fnConv.getSILResultType(IGF.IGM.getMaximalTypeExpansionContext());
+    auto &nativeSchema =
+        IGF.IGM.getTypeInfo(resultType).nativeReturnValueSchema(IGF.IGM);
+    auto expectedNativeResultType = nativeSchema.getExpandedType(IGF.IGM);
+    if (expectedNativeResultType->isVoidTy()) {
+      // If the async return is void, there is no return to move out of the
+      // argument buffer.
+      return;
+    }
+    assert(call->arg_size() == 1);
+    auto context = call->arg_begin()->get();
+    // Gather the values.
+    Explosion nativeExplosion;
+    auto layout = getAsyncContextLayout(IGF, getCallee().getOrigFunctionType(),
+                                        getCallee().getSubstFunctionType(),
+                                        getCallee().getSubstitutions());
+    auto dataAddr = layout.emitCastTo(IGF, context);
+    int index = layout.getFirstDirectReturnIndex();
+    for (auto result :
+         getCallee().getSubstFunctionType()->getDirectFormalResults()) {
+      auto &fieldLayout = layout.getElement(index);
+      Address fieldAddr =
+          fieldLayout.project(IGF, dataAddr, /*offsets*/ llvm::None);
+      auto &fieldTI = fieldLayout.getType();
+      cast<LoadableTypeInfo>(fieldTI).loadAsCopy(IGF, fieldAddr,
```

Comment:
This should be loadAsTake (+0 load). Any return calling convention ownership operation should have happened as part of SIL emission i.e it is not the `return %owned_value : $SomeClass` that performs the +1. Rather it is going to be some `%owned_value = copy_value ` or some other instructions before the return.

We try to use the SILFunctionConventions abstraction rather the directly operating on SILFunctionType API. I would rather use the `SILFunctionConventions(getCallee().getSubstFunctionType(), Module).getDirectSILResults())` from above's `fnConv`  here. Sorry for not noticing/forgetting that when we looked at the code yesterday.

Reply:
Thanks for the explanation!  Fixed.